lib/ruby-mpd/plugins/command_list.rb

```diff
@@ -0,0 +1,177 @@
+module MPD::Plugins
+
+  # Batch send multiple commands at once for speed.
+  module CommandList
+    # Send multiple commands at once.
+    #
+    # By default, any response from the server is ignored (for speed).
+    # To get results, pass +response_type+ as one of the following options:
+    #
+    # * +:hash+   — a single hash of return values; multiple return values for the same key are grouped in an array
+    # * +:values+ — an array of individual values
+    # * +:hashes+ — an array of hashes (where value boundaries are guessed based on the first result)
+    # * +:songs+  — an array of Song instances from the results
+    # * +:playlists+ - an array of Playlist instances from the results
+    #
+    # Note that each supported command has no return value inside the block.
+    # Instead, the block itself returns the array of results.
+    #
+    # @param [Symbol] response_type the type of responses desired.
+    # @return [nil] default behavior.
+    # @return [Array] if +response_type+ is +:values+, +:hashes+, +:songs+, or +:playlists+.
+    # @return [Hash] if +response_type+ is +:hash+.
+    #
+    # @example Simple batched control commands
+    #   @mpd.command_list do
+    #     stop
+    #     shuffle
+    #     save "shuffled"
+    #   end
+    #
+    # @example Adding songs to the queue, ignoring the response
+    #   @mpd.command_list do
+    #     my_songs.each do |song|
+    #       add(song)
+    #     end
+    #   end
+    #
+    # @example Adding songs to the queue and getting the song ids
+    #   ids = @mpd.command_list(:values){ my_songs.each{ |song| addid(song) } }
+    #   #=> [799,800,801,802,803]
+    #
+    #   ids = @mpd.command_list(:hashes){ my_songs.each{ |song| addid(song) } }
+    #   #=> [ {:id=>809}, {:id=>810}, {:id=>811}, {:id=>812}, {:id=>813} ]
+    #
+    #   ids = @mpd.command_list(:hash){   my_songs.each{ |song| addid(song) } }
+    #   #=> { :id=>[804,805,806,807,808] }
+    #
+    # @example Finding songs matching various genres
+    #   songs = @mpd.command_list(:songs) do
+    #     where genre:'Alternative Rock'
+    #     where genre:'Alt. Rock'
+    #     where genre:'alt-rock'
+    #   end
+    #
+    # @see CommandList::Commands CommandList::Commands for a list of supported commands.
+    def command_list(response_type=nil,&commands)
+      @mutex.synchronize do
+        begin
+          socket.puts "command_list_begin"
+          CommandList::Commands.new(self).instance_eval(&commands)
+          socket.puts "command_list_end"
+
+          # clear the response from the socket, even if we will not parse it
+          response = handle_server_response || ""
+
+          case response_type
```

Comment:
Yes, when `response_type` is `nil` (or omitted) I do just throw away the responses.

As I mentioned in another comment, about half my current uses of `command_list` I do use the response. For example, getting `MPD::Song` instances related to a bunch of different URIs:

```ruby
songs = @mpd.command_list(:songs){ uris.each{ |uri| where({file:uri},{strict:true}) } }
```

Yes, the command_list does return multiple responses, and in the 'common' (as I'm imagining them) use cases they will be the same type. There are edge cases where the user might want to send mixed commands with varying return types AND want the return types. That's not currently possible with my current code, since I'm using `command_list_begin` (so all responses come one after another, not separated by "OK") and thus using the existing parsing heuristics to split them up.

I decided to use `command_list_begin` instead of `command_list_ok_begin` to remove all the extra "OK" responses when I don't care about the responses, or when they are short (such as with `addid`. However, I could imagine a different approach that works like this:

* If the user indicates no responses needed, use `command_list_begin` and throw away the response.
* If the user indicates responses needed:
  * Set an instance variable flag that we're in a command list.
  * Use `command_list_ok_begin` so all responses are separated by "list_OK".
  * Create an empty array of response types.
  * When `send_command` is called within the command list, for each command add the response type to the array.
  * Create a new `parse_command_list_response` method in parser (not re-using `parse_response`) that splits the response on "list_OK" and then uses the array of response types to create the appropriate objects.

This would be a little more code than I currently have, and the resulting responses and parsing MIGHT be a little slower, but it would be more DWIM, more robust, and less geeky. Something like:

```ruby
results = mpd.command_list results:true do
  addid(song1)
  where genre:'rock'
  added(song2)
end

p results[0] #=> 42
p results[1] #=> [ <MPD::Song#...>, <MPD::Song#...>, <MPD::Song#...>, ... ]
p results[2] #=> 43
```

I'm not sure if I believe there is a strong need for the intermingling, but I am a fan of things just working as expected (POLS and all that).

Does this sound like a good idea to you?